proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java

```diff
@@ -147,6 +145,11 @@ public boolean handle(PluginMessage packet) {
             player.getConnectedServer().getConnection().write(PluginMessageUtil.rewriteMCBrand(packet));
         } else if (player.getConnectedServer().isLegacyForge() && !player.getConnectedServer().hasCompletedJoin()) {
             if (packet.getChannel().equals(VelocityConstants.FORGE_LEGACY_HANDSHAKE_CHANNEL)) {
+                List<ModInfo.Mod> mods = PluginMessageUtil.readModList(packet);
```

Comment:
Once you have the mod list, there is no need to do it again as the mod list won't change for FML clients during the lifetime of their connection. Rather than go through the trouble of reading _every_ `FML|HS` packet and the overhead that brings, I suggest checking to see if the mod list has been set, and once it has, not doing all this.

Note that if it later transpires that we need to track the FML handshake, this can be incorporated into that.